File: src/material.rs
use crate::config::Float;
use crate::vec3;
use vec3::{Vec3, Color};
use crate::ray::{Ray,Hit};

fn reflect(v_in: &Vec3, normal: &Vec3) -> Vec3
{
    *v_in - 2.0 * vec3::dot(v_in, normal) * *normal
}

// If total reflection, return None.
fn refract(v_in: &Vec3, normal: &Vec3, ni_over_nt: Float) -> Option<Vec3>
{
    let uv = Vec3::unit(v_in);
    let dt = vec3::dot(&uv, normal);
    let discriminant = 1.0 - ni_over_nt * ni_over_nt * (1.0 - dt * dt);
    if discriminant > 0.0
    {
        Some(ni_over_nt * (uv - (*normal)*dt) - (*normal) * discriminant.sqrt())
    }
    else
    {
        None
    }
}

fn schlick(cos: Float, ref_idx: Float) -> Float
{
    let r0: Float = (1.0 - ref_idx) / (1.0 + ref_idx);
    let r02 = r0 * r0;
    r02 + (1.0 - r02) * (1.0 - cos).powi(5)
}

pub trait Material
{
    // Return (ray, attenuation).
    fn scatter(&self, r_in: &Ray, hit: &Hit) -> Option<(Ray, Vec3)>;
}

pub struct Lambertian
{
    pub albedo: Vec3,
}

impl Material for Lambertian
{
    fn scatter(&self, r_in: &Ray, hit: &Hit) -> Option<(Ray, Vec3)>
    {
        let target = hit.p + hit.normal + Vec3::randInUnitSphere();
        Some((Ray { origin: hit.p, dir: target - hit.p }, self.albedo))
    }
}

pub struct Metal
{
    pub albedo: Vec3,
    pub roughness: Float,
}

impl Material for Metal
{
    fn scatter(&self, r_in: &Ray, hit: &Hit) -> Option<(Ray, Vec3)>
    {
        let reflected = reflect(&(Vec3::unit(&r_in.dir)), &hit.normal);
        let scattered = Ray
        {
            origin: hit.p,
            dir: self.roughness * Vec3::randInUnitSphere() + reflected,
        };

        if vec3::dot(&scattered.dir, &hit.normal) > 0.0
        {
            return Some((scattered, self.albedo));
        }
        else
        {
            return None;
        }
    }
}

pub struct Glass
{
    pub ref_index: Float,
}

impl Material for Glass
{
    fn scatter(&self, r_in: &Ray, hit: &Hit) -> Option<(Ray, Vec3)>
    {
        let mut ref_normal = Vec3::origin();
        let mut ni_over_nt: Float = 0.0;
        let attenuation = Vec3::new(1.0, 1.0, 1.0);
        let mut cos: Float;

        if vec3::dot(&r_in.dir, &hit.normal) > 0.0
        {
            ref_normal = -(hit.normal);
            ni_over_nt = self.ref_index;
            cos = self.ref_index * vec3::dot(&r_in.dir, &hit.normal) / r_in.dir.norm();
        }
        else
        {
            ref_normal = hit.normal;
            ni_over_nt = 1.0 / self.ref_index;
            cos = -vec3::dot(&r_in.dir, &hit.normal) / r_in.dir.norm();
        }

        let mut reflect_prob = schlick(cos, self.ref_index);
        if let Some(refracted) = refract(&r_in.dir, &ref_normal, ni_over_nt)
        {
            if rand::random::<Float>() < reflect_prob
            {
                let reflected = reflect(&r_in.dir, &hit.normal);
                return Some((Ray{origin: hit.p, dir: reflected}, attenuation));
            }
            else
            {
                return Some((Ray{origin: hit.p, dir: refracted}, attenuation));
            }
        }
        else
        {
            let reflected = reflect(&r_in.dir, &hit.normal);
            return Some((Ray{origin: hit.p, dir: reflected}, attenuation));
        }
    }
}

pub struct Null {}

impl Material for Null
{
    fn scatter(&self, r_in: &Ray, hit: &Hit) -> Option<(Ray, Vec3)>
    {
        None
    }
}

pub static NULL: Null = Null{};

File: cpp/src/material.cpp
#include "material.h"
#include "utils.h"

namespace lorentz
{
    Vec3 reflect(const Vec3& in, const Vec3& normal)
    {
        return in - 2.0f * Vec3::dot(in, normal) * normal;
    }

    std::optional<Vec3> refract(const Vec3& in, const Vec3& normal,
                                const Float ni_over_nt)
    {
        auto uv = Vec3::unit(in);
        auto dt = Vec3::dot(uv, normal);
        auto discriminant = 1.0f - ni_over_nt * ni_over_nt * (1.0f - dt * dt);
        if(discriminant > 0.0f)
        {
            return ni_over_nt * (uv - normal*dt) - normal * std::sqrt(discriminant);
        }
        else
        {
            return std::nullopt;
        }
    }

    Float schlick(const Float cos, const Float ref_idx)
    {
        auto r0 = (1.0f - ref_idx) / (1.0f + ref_idx);
        auto r02 = r0 * r0;
        return r02 + (1.0f - r02) * std::pow(1.0f - cos, 5);
    }


    std::optional<std::pair<Ray, Vec3>>
    Lambertian :: scatter(const Ray& in, const Hit& hit) const
    {
        UNUSED(in);
        auto target = hit.p + hit.normal + Vec3::randInUnitSphere();
        return std::make_pair(Ray{ hit.p, target - hit.p },
                              albedo);
    }

    LambertianRandomColor :: LambertianRandomColor()
    {
        albedo = randomColor();
    }

    Color LambertianRandomColor :: randomColor() const
    {
        return Color(random(0.1, 0.7), random(0.1, 0.7), random(0.1, 0.7));
    }

    std::optional<std::pair<Ray, Vec3>>
    Metal :: scatter(const Ray& in, const Hit& hit) const
    {
        auto reflected = reflect(Vec3::unit(in.dir), hit.normal);
        auto scattered = Ray{
            hit.p,
            roughness * Vec3::randInUnitSphere() + reflected,
        };

        if(Vec3::dot(scattered.dir, hit.normal) > 0.0f)
        {
            return std::make_pair(scattered, albedo);
        }
        else
        {
            return std::nullopt;
        }
    }

    std::optional<std::pair<Ray, Vec3>>
    Glass :: scatter(const Ray& in, const Hit& hit) const
    {
        Vec3 ref_normal = Vec3::origin();
        Float ni_over_nt = 0.0f;
        Vec3 attenuation = Vec3(1.0f, 1.0f, 1.0f);
        Float cos = 0.0f;

        if(Vec3::dot(in.dir, hit.normal) > 0.0f)
        {
            ref_normal = -(hit.normal);
            ni_over_nt = ref_index;
            cos = ref_index * Vec3::dot(in.dir, hit.normal) / in.dir.norm();
        }
        else
        {
            ref_normal = hit.normal;
            ni_over_nt = 1.0f / ref_index;
            cos = -Vec3::dot(in.dir, hit.normal) / in.dir.norm();
        }

        auto reflect_prob = schlick(cos, ref_index);
        auto refracted = refract(in.dir, ref_normal, ni_over_nt);
        if(refracted.has_value())
        {
            if(random() < reflect_prob)
            {
                auto reflected = reflect(in.dir, hit.normal);
                return std::make_pair(Ray{hit.p, reflected}, attenuation);
            }
            else
            {
                return std::make_pair(Ray{hit.p, *refracted}, attenuation);
            }
        }
        else
        {
            auto reflected = reflect(in.dir, hit.normal);
            return std::make_pair(Ray{hit.p, reflected}, attenuation);
        }
    }
} // namespace lorentz
